lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -1302,6 +1304,278 @@ void ESItoHWPass::runOnOperation() {
     signalPassFailure();
 }
 
+//===----------------------------------------------------------------------===//
+// Wire up services pass.
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Implements a pass to connect up ESI services clients to the nearest server
+/// instantiation. Wires up the ports and generates a generation request to call
+/// a user-specified generator.
+struct ESIConnectServicesPass
+    : public ESIConnectServicesBase<ESIConnectServicesPass>,
+      msft::PassCommon {
+  void runOnOperation() override;
+
+  /// "Bubble up" the specified requests to all of the instantiations of the
+  /// module specified. Create and connect up ports to tunnel the ESI channels
+  /// through.
+  LogicalResult surfaceReqs(hw::HWModuleLike,
+                            ArrayRef<RequestToClientConnection>,
+                            ArrayRef<RequestToServerConnection>);
+  /// For any service which is "local" (provides the requested service) in a
+  /// module, replace it with a ServiceImplementOp. Said op is to be replaced
+  /// with an instantiation by a generator.
+  LogicalResult replaceInst(ServiceInstanceOp, Block *portReqs);
+  /// Figure out which requests are "local" vs need to be surfaced. Call
+  /// 'surfaceReqs' and/or 'replaceInst' as appropriate.
+  LogicalResult process(hw::HWModuleLike);
+};
+} // anonymous namespace
+
+void ESIConnectServicesPass::runOnOperation() {
+  ModuleOp outerMod = getOperation();
+  topLevelSyms.addDefinitions(outerMod);
+  if (failed(verifyInstances(outerMod))) {
+    signalPassFailure();
+    return;
+  }
+
+  // Get a partially-ordered list of modules based on the instantiation DAG.
+  SmallVector<hw::HWModuleLike, 64> sortedMods;
+  getAndSortModules(outerMod, sortedMods);
```

Comment:
I think a comment here should state the importance of iterating over the instance hierarchy in post-order (which is what `getAndSortModules` returns) instead of the rather vague 'partially-ordered list'. If we didn't do a bottom-up traversal of the instance hierarchy, we'd risk rewriting module interfaces multiple times during traversal (correct, but slow).

Reply:
It's actually worse than that...